dev/usb/class/cdcserial/cdcserial.c

```diff
@@ -0,0 +1,331 @@
+ * publish, distribute, sublicense, and/or sell copies of the Software,
+ * and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <dev/usb/class/cdcserial.h>
+
+#include <assert.h>
+#include <dev/udc.h>
+#include <dev/usb.h>
+#include <dev/usbc.h>
+#include <dev/usbc.h>
+#include <err.h>
+#include <kernel/event.h>
+#include <sys/types.h>
+#include <trace.h>
+
+#define LOCAL_TRACE 0
+
+#define W(w) (w & 0xff), (w >> 8)
+#define W3(w) (w & 0xff), ((w >> 8) & 0xff), ((w >> 16) & 0xff)
+
+#define CDC_REQ_SEND_CMD               0x00
+#define CDC_REQ_GET_RESP               0x01
+#define CDC_REQ_SET_COMM_FEATURE       0x02
+#define CDC_REQ_GET_COMM_FEATURE       0x03
+#define CDC_REQ_CLEAR_COMM_FEATURE     0x04
+#define CDC_REQ_SET_LINE_CODING        0x20
+#define CDC_REQ_GET_LINE_CODING        0x21
+#define CDC_REQ_SET_CONTROL_LINE_STATE 0x22
+#define CDC_REQ_SEND_BREAK             0x23
+
+#define EP0_MTU (64)
+static event_t txevt = EVENT_INITIAL_VALUE(txevt, 0, EVENT_FLAG_AUTOUNSIGNAL);
+static event_t rxevt = EVENT_INITIAL_VALUE(rxevt, 0, EVENT_FLAG_AUTOUNSIGNAL);
+
+static volatile bool usb_online = false;
+
+/* top level device descriptor */
+static const uint8_t dev_descr[] = {
+    0x12,           /* descriptor length */
+    DEVICE,         /* Device Descriptor type */
+    W(0x0200),      /* USB Version */
+    0x00,           /* class: infer from interfaces */
+    0x00,           /* subclass: infer from interfaces */
+    0x00,           /* protocol infer from interfaces */
+    64,             /* max packet size for endpoint 0 */
+    W(0x9999),      /* vendor */
+    W(0x9999),      /* product */
+    W(0x9999),      /* release */
+    0x2,            /* manufacturer string */
+    0x1,            /* product string */
+    0x0,            /* serialno string */
+    0x1,            /* num configs */
+};
+
+/* high/low speed device qualifier */
+static const uint8_t devqual_descr[] = {
+    0x0a,             /* descriptor length */
+    DEVICE_QUALIFIER, /* Device Qualifier type */
+    W(0x0200),        /* USB version */
+    0x00,             /* class */
+    0x00,             /* subclass */
+    0x00,             /* protocol */
+    64,               /* max packet size for endpoint 0 */
```

Comment:
Actually no. It's the transfer size of a single usb packet. 0x40 is the maximum size for a control transfer. Technically this should come from the configuration of the device controller, but no real device supports less than the max.